File: core/src/test/scala/unit/kafka/tools/ConsoleProducerTest.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package kafka.tools

import kafka.common.MessageReader

import kafka.tools.ConsoleProducer.LineMessageReader
import kafka.utils.{Exit, TestUtils}
import org.apache.kafka.clients.producer.{Producer, ProducerConfig, ProducerRecord}
import org.apache.kafka.tools.api.RecordReader
import org.junit.jupiter.api.Assertions.{assertEquals, assertThrows, assertTrue}
import org.junit.jupiter.api.Test
import org.mockito.Mockito

import java.io.InputStream
import java.util
import java.util.Properties
import scala.annotation.nowarn

class ConsoleProducerTest {

  val brokerListValidArgs: Array[String] = Array(
    "--broker-list",
    "localhost:1001,localhost:1002",
    "--topic",
    "t3",
    "--property",
    "parse.key=true",
    "--property",
    "key.separator=#"
  )
  val bootstrapServerValidArgs: Array[String] = Array(
    "--bootstrap-server",
    "localhost:1003,localhost:1004",
    "--topic",
    "t3",
    "--property",
    "parse.key=true",
    "--property",
    "key.separator=#"
  )
  val invalidArgs: Array[String] = Array(
    "--t", // not a valid argument
    "t3"
  )
  val bootstrapServerOverride: Array[String] = Array(
    "--broker-list",
    "localhost:1001",
    "--bootstrap-server",
    "localhost:1002",
    "--topic",
    "t3",
  )
  val clientIdOverride: Array[String] = Array(
    "--broker-list",
    "localhost:1001",
    "--topic",
    "t3",
    "--producer-property",
    "client.id=producer-1"
  )
  val batchSizeOverriddenByMaxPartitionMemoryBytesValue: Array[String] = Array(
    "--broker-list",
    "localhost:1001",
    "--bootstrap-server",
    "localhost:1002",
    "--topic",
    "t3",
    "--batch-size",
    "123",
    "--max-partition-memory-bytes",
    "456"
  )
  val btchSizeSetAndMaxPartitionMemoryBytesNotSet: Array[String] = Array(
    "--broker-list",
    "localhost:1001",
    "--bootstrap-server",
    "localhost:1002",
    "--topic",
    "t3",
    "--batch-size",
    "123"
  )
  val batchSizeNotSetAndMaxPartitionMemoryBytesSet: Array[String] = Array(
    "--broker-list",
    "localhost:1001",
    "--bootstrap-server",
    "localhost:1002",
    "--topic",
    "t3",
    "--max-partition-memory-bytes",
    "456"
  )
  val batchSizeDefault: Array[String] = Array(
    "--broker-list",
    "localhost:1001",
    "--bootstrap-server",
    "localhost:1002",
    "--topic",
    "t3"
  )

  @Test
  def testValidConfigsBrokerList(): Unit = {
    val config = new ConsoleProducer.ProducerConfig(brokerListValidArgs)
    val producerConfig = new ProducerConfig(ConsoleProducer.producerProps(config))
    assertEquals(util.Arrays.asList("localhost:1001", "localhost:1002"),
      producerConfig.getList(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG))
  }

  @Test
  def testValidConfigsBootstrapServer(): Unit = {
    val config = new ConsoleProducer.ProducerConfig(bootstrapServerValidArgs)
    val producerConfig = new ProducerConfig(ConsoleProducer.producerProps(config))
    assertEquals(util.Arrays.asList("localhost:1003", "localhost:1004"),
      producerConfig.getList(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG))
  }

  @Test
  def testInvalidConfigs(): Unit = {
    Exit.setExitProcedure((_, message) => throw new IllegalArgumentException(message.orNull))
    try assertThrows(classOf[IllegalArgumentException], () => new ConsoleProducer.ProducerConfig(invalidArgs))
    finally Exit.resetExitProcedure()
  }

  @Test
  def testParseKeyProp(): Unit = {
    val config = new ConsoleProducer.ProducerConfig(brokerListValidArgs)
    val reader = Class.forName(config.readerClass).getDeclaredConstructor().newInstance().asInstanceOf[LineMessageReader]
    reader.configure(ConsoleProducer.getReaderProps(config).asInstanceOf[java.util.Map[String, _]])
    assertTrue(reader.keySeparator == "#")
    assertTrue(reader.parseKey)
  }

  @Test
  def testParseReaderConfigFile(): Unit = {
    val propsFile = TestUtils.tempPropertiesFile(Map("parse.key" -> "true", "key.separator" -> "|"))

    val args = Array(
      "--bootstrap-server", "localhost:9092",
      "--topic", "test",
      "--property", "key.separator=;",
      "--property", "parse.headers=true",
      "--reader-config", propsFile.getAbsolutePath
    )
    val config = new ConsoleProducer.ProducerConfig(args)
    val reader = Class.forName(config.readerClass).getDeclaredConstructor().newInstance().asInstanceOf[LineMessageReader]
    reader.configure(ConsoleProducer.getReaderProps(config).asInstanceOf[java.util.Map[String, _]])
    assertEquals(";", reader.keySeparator)
    assertTrue(reader.parseKey)
    assertTrue(reader.parseHeaders)
  }

  @Test
  def testBootstrapServerOverride(): Unit = {
    val config = new ConsoleProducer.ProducerConfig(bootstrapServerOverride)
    val producerConfig = new ProducerConfig(ConsoleProducer.producerProps(config))
    assertEquals(util.Arrays.asList("localhost:1002"),
      producerConfig.getList(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG))
  }

  @Test
  def testClientIdOverride(): Unit = {
    val config = new ConsoleProducer.ProducerConfig(clientIdOverride)
    val producerConfig = new ProducerConfig(ConsoleProducer.producerProps(config))
    assertEquals("producer-1",
      producerConfig.getString(ProducerConfig.CLIENT_ID_CONFIG))
  }

  @Test
  def testDefaultClientId(): Unit = {
    val config = new ConsoleProducer.ProducerConfig(brokerListValidArgs)
    val producerConfig = new ProducerConfig(ConsoleProducer.producerProps(config))
    assertEquals("console-producer",
      producerConfig.getString(ProducerConfig.CLIENT_ID_CONFIG))
  }

  @Test
  def testBatchSizeOverriddenByMaxPartitionMemoryBytesValue(): Unit = {
    val config = new ConsoleProducer.ProducerConfig(batchSizeOverriddenByMaxPartitionMemoryBytesValue)
    val producerConfig = new ProducerConfig(ConsoleProducer.producerProps(config))
    assertEquals(456,
      producerConfig.getInt(ProducerConfig.BATCH_SIZE_CONFIG))
  }

  @Test
  def testBatchSizeSetAndMaxPartitionMemoryBytesNotSet(): Unit = {
    val config = new ConsoleProducer.ProducerConfig(btchSizeSetAndMaxPartitionMemoryBytesNotSet)
    val producerConfig = new ProducerConfig(ConsoleProducer.producerProps(config))
    assertEquals(123,
      producerConfig.getInt(ProducerConfig.BATCH_SIZE_CONFIG))
  }

  @Test
  def testDefaultBatchSize(): Unit = {
    val config = new ConsoleProducer.ProducerConfig(batchSizeDefault)
    val producerConfig = new ProducerConfig(ConsoleProducer.producerProps(config))
    assertEquals(16*1024,
      producerConfig.getInt(ProducerConfig.BATCH_SIZE_CONFIG))
  }

  @Test
  def testBatchSizeNotSetAndMaxPartitionMemoryBytesSet (): Unit = {
    val config = new ConsoleProducer.ProducerConfig(batchSizeNotSetAndMaxPartitionMemoryBytesSet)
    val producerConfig = new ProducerConfig(ConsoleProducer.producerProps(config))
    assertEquals(456,
      producerConfig.getInt(ProducerConfig.BATCH_SIZE_CONFIG))
  }

  @Test
  def testNewReader(): Unit = {
    ConsoleProducerTest.configureCount = 0
    ConsoleProducerTest.closeCount = 0
    val reader = ConsoleProducer.newReader(classOf[ConsoleProducerTest.TestMessageReader].getName, new Properties())
    // the deprecated MessageReader get configured when creating records
    assertEquals(0, ConsoleProducerTest.configureCount)
    reader.readRecords(System.in)
    assertEquals(1, ConsoleProducerTest.configureCount)
    assertEquals(0, ConsoleProducerTest.closeCount)
    assertThrows(classOf[IllegalStateException], () => reader.readRecords(System.in))
    reader.close()
    assertEquals(1, ConsoleProducerTest.closeCount)

    ConsoleProducerTest.configureCount = 0
    ConsoleProducerTest.closeCount = 0

    val reader1 = ConsoleProducer.newReader(classOf[ConsoleProducerTest.TestRecordReader].getName, new Properties())
    assertEquals(1, ConsoleProducerTest.configureCount)
    assertEquals(0, ConsoleProducerTest.closeCount)
    reader1.close()
    assertEquals(1, ConsoleProducerTest.closeCount)
  }

  @Test
  def testLoopReader(): Unit = {
    ConsoleProducerTest.configureCount = 0
    ConsoleProducerTest.closeCount = 0
    val reader = ConsoleProducer.newReader(classOf[ConsoleProducerTest.TestRecordReader].getName, new Properties())

    ConsoleProducer.loopReader(Mockito.mock(classOf[Producer[Array[Byte], Array[Byte]]]),
      reader, System.in, false)

    assertEquals(1, ConsoleProducerTest.configureCount)
    assertEquals(1, ConsoleProducerTest.closeCount)
  }
}

@nowarn("cat=deprecation")
object ConsoleProducerTest {
  var configureCount = 0
  var closeCount = 0
  class TestMessageReader extends MessageReader {
    override def init(inputStream: InputStream, props: Properties): Unit = configureCount += 1
    override def readMessage(): ProducerRecord[Array[Byte], Array[Byte]] = null

    override def close(): Unit = closeCount += 1

  }

  class TestRecordReader extends RecordReader {
    override def configure(configs: util.Map[String, _]): Unit = configureCount += 1
    override def readRecords(inputStream: InputStream): java.util.Iterator[ProducerRecord[Array[Byte], Array[Byte]]] =
      java.util.Collections.emptyIterator()

    override def close(): Unit = closeCount += 1
  }
}

File: tools/src/test/java/org/apache/kafka/tools/ConsoleProducerTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.kafka.tools;

import kafka.utils.TestUtils;

import org.apache.kafka.clients.producer.Producer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.utils.Exit;
import org.apache.kafka.tools.ConsoleProducer.ConsoleProducerOptions;
import org.apache.kafka.tools.api.RecordReader;

import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.file.Files;
import java.util.Collections;
import java.util.Iterator;
import java.util.Map;

import static java.util.Arrays.asList;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class ConsoleProducerTest {
    private static final String[] BOOTSTRAP_SERVER_VALID_ARGS = new String[]{
        "--bootstrap-server", "localhost:1003,localhost:1004",
        "--topic", "t3",
        "--property", "parse.key=true",
        "--property", "key.separator=#"
    };
    private static final String[] INVALID_ARGS = new String[]{
        "--t", // not a valid argument
        "t3"
    };
    private static final String[] BOOTSTRAP_SERVER_OVERRIDE = new String[]{
        "--bootstrap-server", "localhost:1002",
        "--topic", "t3",
    };
    private static final String[] CLIENT_ID_OVERRIDE = new String[]{
        "--bootstrap-server", "localhost:1001",
        "--topic", "t3",
        "--producer-property", "client.id=producer-1"
    };
    private static final String[] BATCH_SIZE_OVERRIDDEN_BY_MAX_PARTITION_MEMORY_BYTES_VALUE = new String[]{
        "--bootstrap-server", "localhost:1002",
        "--topic", "t3",
        "--batch-size", "123",
        "--max-partition-memory-bytes", "456"
    };
    private static final String[] BATCH_SIZE_SET_AND_MAX_PARTITION_MEMORY_BYTES_NOT_SET = new String[]{
        "--bootstrap-server", "localhost:1002",
        "--topic", "t3",
        "--batch-size", "123"
    };
    private static final String[] BATCH_SIZE_NOT_SET_AND_MAX_PARTITION_MEMORY_BYTES_SET = new String[]{
        "--bootstrap-server", "localhost:1002",
        "--topic", "t3",
        "--max-partition-memory-bytes", "456"
    };
    private static final String[] BATCH_SIZE_DEFAULT = new String[]{
        "--bootstrap-server", "localhost:1002",
        "--topic", "t3"
    };
    private static final String[] TEST_RECORD_READER = new String[]{
        "--bootstrap-server", "localhost:1002",
        "--topic", "t3",
        "--line-reader", TestRecordReader.class.getName()
    };

    @Test
    public void testValidConfigsBootstrapServer() throws IOException {
        ConsoleProducerOptions opts = new ConsoleProducerOptions(BOOTSTRAP_SERVER_VALID_ARGS);
        ProducerConfig producerConfig = new ProducerConfig(opts.producerProps());

        assertEquals(asList("localhost:1003", "localhost:1004"), producerConfig.getList(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG));
    }

    @Test
    public void testInvalidConfigs() {
        Exit.setExitProcedure((statusCode, message) -> {
            throw new IllegalArgumentException(message);
        });
        try {
            assertThrows(IllegalArgumentException.class, () -> new ConsoleProducerOptions(INVALID_ARGS));
        } finally {
            Exit.resetExitProcedure();
        }
    }

    @Test
    public void testParseKeyProp() throws ReflectiveOperationException, IOException {
        ConsoleProducerOptions opts = new ConsoleProducerOptions(BOOTSTRAP_SERVER_VALID_ARGS);
        LineMessageReader reader = (LineMessageReader) Class.forName(opts.readerClass()).getDeclaredConstructor().newInstance();
        reader.configure(opts.readerProps());

        assertEquals("#", reader.keySeparator());
        assertTrue(reader.parseKey());
    }

    @Test
    public void testParseReaderConfigFile() throws Exception {
        File propsFile = TestUtils.tempFile();
        OutputStream propsStream = Files.newOutputStream(propsFile.toPath());
        propsStream.write("parse.key=true\n".getBytes());
        propsStream.write("key.separator=|".getBytes());
        propsStream.close();

        String[] args = new String[]{
            "--bootstrap-server", "localhost:9092",
            "--topic", "test",
            "--property", "key.separator=;",
            "--property", "parse.headers=true",
            "--reader-config", propsFile.getAbsolutePath()
        };
        ConsoleProducerOptions opts = new ConsoleProducerOptions(args);
        LineMessageReader reader = (LineMessageReader) Class.forName(opts.readerClass()).getDeclaredConstructor().newInstance();
        reader.configure(opts.readerProps());

        assertEquals(";", reader.keySeparator());
        assertTrue(reader.parseKey());
        assertTrue(reader.parseHeaders());
    }

    @Test
    public void testBootstrapServerOverride() throws IOException {
        ConsoleProducerOptions opts = new ConsoleProducerOptions(BOOTSTRAP_SERVER_OVERRIDE);
        ProducerConfig producerConfig = new ProducerConfig(opts.producerProps());

        assertEquals(Collections.singletonList("localhost:1002"), producerConfig.getList(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG));
    }

    @Test
    public void testClientIdOverride() throws IOException {
        ConsoleProducerOptions opts = new ConsoleProducerOptions(CLIENT_ID_OVERRIDE);
        ProducerConfig producerConfig = new ProducerConfig(opts.producerProps());

        assertEquals("producer-1", producerConfig.getString(ProducerConfig.CLIENT_ID_CONFIG));
    }

    @Test
    public void testDefaultClientId() throws IOException {
        ConsoleProducerOptions opts = new ConsoleProducerOptions(BOOTSTRAP_SERVER_VALID_ARGS);
        ProducerConfig producerConfig = new ProducerConfig(opts.producerProps());

        assertEquals("console-producer", producerConfig.getString(ProducerConfig.CLIENT_ID_CONFIG));
    }

    @Test
    public void testBatchSizeOverriddenByMaxPartitionMemoryBytesValue() throws IOException {
        ConsoleProducerOptions opts = new ConsoleProducerOptions(BATCH_SIZE_OVERRIDDEN_BY_MAX_PARTITION_MEMORY_BYTES_VALUE);
        ProducerConfig producerConfig = new ProducerConfig(opts.producerProps());

        assertEquals(456, producerConfig.getInt(ProducerConfig.BATCH_SIZE_CONFIG));
    }

    @Test
    public void testBatchSizeSetAndMaxPartitionMemoryBytesNotSet() throws IOException {
        ConsoleProducerOptions opts = new ConsoleProducerOptions(BATCH_SIZE_SET_AND_MAX_PARTITION_MEMORY_BYTES_NOT_SET);
        ProducerConfig producerConfig = new ProducerConfig(opts.producerProps());

        assertEquals(123, producerConfig.getInt(ProducerConfig.BATCH_SIZE_CONFIG));
    }

    @Test
    public void testDefaultBatchSize() throws IOException {
        ConsoleProducerOptions opts = new ConsoleProducerOptions(BATCH_SIZE_DEFAULT);
        ProducerConfig producerConfig = new ProducerConfig(opts.producerProps());

        assertEquals(16 * 1024, producerConfig.getInt(ProducerConfig.BATCH_SIZE_CONFIG));
    }

    @Test
    public void testBatchSizeNotSetAndMaxPartitionMemoryBytesSet() throws IOException {
        ConsoleProducerOptions opts = new ConsoleProducerOptions(BATCH_SIZE_NOT_SET_AND_MAX_PARTITION_MEMORY_BYTES_SET);
        ProducerConfig producerConfig = new ProducerConfig(opts.producerProps());

        assertEquals(456, producerConfig.getInt(ProducerConfig.BATCH_SIZE_CONFIG));
    }

    @Test
    public void testNewReader() throws Exception {
        ConsoleProducer producer = new ConsoleProducer();
        TestRecordReader reader = (TestRecordReader) producer.messageReader(new ConsoleProducerOptions(TEST_RECORD_READER));

        assertEquals(1, reader.configureCount());
        assertEquals(0, reader.closeCount());

        reader.close();
        assertEquals(1, reader.closeCount());
    }

    @Test
    @SuppressWarnings("unchecked")
    public void testLoopReader() throws Exception {
        ConsoleProducer producer = new ConsoleProducer();
        TestRecordReader reader = (TestRecordReader) producer.messageReader(new ConsoleProducerOptions(TEST_RECORD_READER));

        producer.loopReader(Mockito.mock(Producer.class), reader, false);

        assertEquals(1, reader.configureCount());
        assertEquals(1, reader.closeCount());
    }

    public static class TestRecordReader implements RecordReader {
        private int configureCount = 0;
        private int closeCount = 0;

        @Override
        public void configure(Map<String, ?> configs) {
            configureCount += 1;
        }

        @Override
        public Iterator<ProducerRecord<byte[], byte[]>> readRecords(InputStream inputStream) {
            return Collections.emptyIterator();
        }

        @Override
        public void close() {
            closeCount += 1;
        }

        public int configureCount() {
            return configureCount;
        }

        public int closeCount() {
            return closeCount;
        }
    }
}
